New: Exception message template for uncaught rule error (refs #5221)

platinumazure opened the pull request:
Hi folks. I'm asking for [30% feedback](http://blog.sandglaz.com/30-percent-feedback-rule/) here (meaning this PR is probably not ready to merge yet). Please take a look and let me know if this is the right direction.

Terminology:

* Rule action: That is my name for the node event handler in rules, i.e., the values returned in the rule creator function.

The highlights:

* When lib/eslint.js calls `api.on(nodeType)`, I am wrapping the rule action function in a containing function which will call the rule action with try/catch, and if an exception is thrown, it will decorate the exception with metadata representing a new message template.
* New message template contains information on how to report the bug. Probably needs some polishing and copy-editing.

Done:

- [x] Get the filename into the error message
- [x] Mention plugin version as an important piece of information for a plugin bug
- [x] Show parser name, distinguish between default and non-default parser

Do we want these for an initial PR?
- [ ] Different top-level message for default vs non-default parser?
- [ ] Report on node metadata (e.g., line and column number)?
- [ ] Crazy thought: Get the RuleContext object into the wrapper, so that we can do `context.getSourceCode().getText(node)`?

Feedback welcome!

platinumazure commented on lib/eslint.js:
```diff
@@ -27,7 +27,8 @@ var lodash = require("lodash"),
     assert = require("assert"),
-    CodePathAnalyzer = require("./code-path-analysis/code-path-analyzer");
+    CodePathAnalyzer = require("./code-path-analysis/code-path-analyzer"),
+    ESLINT_VERSION = require("../package.json").version;
```
Let me know if there's a better way to do this.

btmills replied:
This is [exactly how we do it](https://github.com/eslint/eslint/blob/7051465f9090959da031b38036673106310bf1be/lib/cli.js#L144) for the `--version` CLI option.

platinumazure commented on lib/eslint.js:
```diff
@@ -505,6 +506,46 @@ function stripUnicodeBOM(text) {
+                ruleName: ruleName,
+                pluginName: pluginName,
+                eslintVersion: ESLINT_VERSION,
+                stackTrace: err.stack
```
I'm already seeing that I never actually use the originalMessage bit. I probably should.

platinumazure commented on messages/rule-crashed.txt:
```diff
@@ -0,0 +1,21 @@
+    You can run this command to get your configuration:
+        eslint --print-config path/to/your/file.js
+
+Be sure that your bug report also clearly calls out any unusual configuration properties, including but not limited to which parser you are using.
```
Obviously this should be cleaned up if I can send the parser metadata into the template.

btmills commented on messages/rule-crashed.txt:
```diff
@@ -0,0 +1,21 @@
+ESLint encountered an uncaught exception from rule "<%- ruleName %>".
+
+<% if (pluginName) { %>
+The rule in question comes from the plugin "eslint-plugin-<%- pluginName %>". You should open an issue on that plugin's issue tracker. Be sure to include your configuration, file source code, and the error message and stack trace.
```
We should either show the plugin's version below next to the ESLint version (if that's possible) or include it in the list of info here.

platinumazure replied:
Good call.

platinumazure commented:
I aim to do some incremental improvements tonight (US time):

* Get filename into error message
* Mention plugin version as an important piece of information for a bug

ilyavolodin commented:
@platinumazure `ruleContext` has filename in it. So you should be able to extract it from there. There is also a reference to the `parser` attached to `ruleContext`, but I think it's the whole executable object, and not just the name (although I might be wrong).

platinumazure commented:
Thanks @ilyavolodin for the tips. I wasn't too worried about the filename since I knew `RuleContext` was constructed around a parsed file, but thanks for saving me a few minutes of digging. And I'll check out the parser as well. (I thought I remembered a `parserName` property floating around somewhere, but that might not be on the `RuleContext` instance.)

ilyavolodin commented:
Actually just checked. `parserName` should also be available on `ruleContext`. It was added in #4649

BYK commented on messages/rule-crashed.txt:
```diff
@@ -0,0 +1,21 @@
+<% if (pluginName) { %>
+The rule in question comes from the plugin "eslint-plugin-<%- pluginName %>". You should open an issue on that plugin's issue tracker. Be sure to include your configuration, file source code, and the error message and stack trace.
+<% } else { %>
+This is a core rule, so this exception may indicate a bug in ESLint. Please open an issue in our issue tracker, using the following information in addition to your exact source code and configuration.
```
Pretty sure we need to check if we are using a custom parser or not.

platinumazure replied:
@BYK Definitely, I just wanted to get something out there and I wasn't sure (at the time) how to detect a custom parser. @ilyavolodin was kind enough to dig up the PR showing where I can find it.

BYK commented:
Great idea btw, thanks for the PR!

platinumazure commented:
Updated with file name, parser name (and whether it represents the default parser or not), and a note about installed plugin version being important.

I've modified my initial post to cover what's done and what's not yet done. I do think it might be worthwhile to limit the scope of this PR to focus on high-value low-hanging fruit, and we can then focus on some of the other enhancements later.

ilyavolodin commented on lib/eslint.js:
```diff
@@ -505,6 +506,53 @@ function stripUnicodeBOM(text) {
+    }
+
+    return function() {
+        try {
```
I don't know of any other way of doing this (other then global error handler), but this will most likely have a major impact on performance. Could you run `npm run perf` before and after this change?

platinumazure commented:
At the request of @ilyavolodin, here are the perf metrics:

* Single-file: 12784.163798ms to 14395.568591ms (12.6% slower)
* Multi-file: 40937.154772ms to 46614.713000ms (13.9% slower)

Yeah, not great.

Since I'm allowing the exception to propagate anyway, there has to be a way for me to put a wrapper around the estraverse traversal instead of the rule actions. I will look into that.

Full dumps:

Before:

> > eslint@2.3.0 perf C:\Users\kpartington\Documents\GitHub\eslint
> > node Makefile.js perf
> 
> 
> Loading:
>   Load performance Run #1:  852.383636ms
>   Load performance Run #2:  861.442609ms
>   Load performance Run #3:  847.120133ms
>   Load performance Run #4:  860.691636ms
>   Load performance Run #5:  867.616382ms
> 
>   Load Performance median:  860.691636ms
> 
> 
> Single File:
>   CPU Speed is 2294 with multiplier 13000000
>   Performance Run #1:  12920.209023ms
>   Performance Run #2:  12696.805085ms
>   Performance Run #3:  12798.272539ms
>   Performance Run #4:  12782.383417000001ms
>   Performance Run #5:  12784.163798ms
> 
>   Performance budget exceeded: 12784.163798ms (limit: 5666.957279860506ms)
> 
> 
> Multi Files (450 files):
>   CPU Speed is 2294 with multiplier 39000000
>   Performance Run #1:  40914.100396ms
>   Performance Run #2:  40983.377754ms
>   Performance Run #3:  40652.639136ms
>   Performance Run #4:  40937.154772ms
>   Performance Run #5:  41096.501573ms
> 
>   Performance budget exceeded: 40937.154772ms (limit: 17000.871839581516ms)

After:

> > eslint@2.3.0 perf C:\Users\kpartington\Documents\GitHub\eslint
> > node Makefile.js perf
> 
> 
> Loading:
>   Load performance Run #1:  882.735344ms
>   Load performance Run #2:  888.435241ms
>   Load performance Run #3:  894.15611ms
>   Load performance Run #4:  892.627838ms
>   Load performance Run #5:  889.052351ms
> 
>   Load Performance median:  889.052351ms
> 
> 
> Single File:
>   CPU Speed is 2294 with multiplier 13000000
>   Performance Run #1:  14395.568591ms
>   Performance Run #2:  13943.606535ms
>   Performance Run #3:  14556.950114ms
>   Performance Run #4:  14166.329555ms
>   Performance Run #5:  15868.433394ms
> 
>   Performance budget exceeded: 14395.568591ms (limit: 5666.957279860506ms)
> 
> 
> Multi Files (450 files):
>   CPU Speed is 2294 with multiplier 39000000
>   Performance Run #1:  49652.101899ms
>   Performance Run #2:  46757.982824ms
>   Performance Run #3:  46517.312312ms
>   Performance Run #4:  46065.083868ms
>   Performance Run #5:  46614.713ms
> 
>   Performance budget exceeded: 46614.713ms (limit: 17000.871839581516ms)

platinumazure commented:
Okay, I've done a quick rewrite to surround `controller.traverse()` with a try/catch.

Perf results:
* Single file: 12844.976982ms (0.5% slower)
* Multi file: 41414.348267ms (1.2% slower)

...And I've already run into the first hurdle. I also lose the name of the rule that threw by going to a more higher-level handler. (I could try parsing it from the error's stack trace, but...) This also means we can't easily tell if the rule was a core rule or plugin rule.

@ilyavolodin What do you think? Do we accept the loss of the rule name and plugin and trust that users (or we) can read a stack trace, or do we accept the performance hit from the original approach and I need to look for ways to offset it?

ilyavolodin commented:
@platinumazure I'm trying to think if this is a right approach in general. Taking perf hit of 1.2% (and it's going to be larger as the number of files increases) for error tracking might be a bit pushing it. Maybe instead we should try to use global error handler and parse out as much information as we can from stack trace? I know we are not going to get nearly as much information, and also it will not be easy to parse stack out, but in that case we will only get hit when the error occurs vs. every time. Not really sure, honestly. @eslint/eslint-team what do you think?

btmills commented:
What if we had an error helper singleton that allowed us to tell it whenever we start lifting another file the name of the file and the config we're using? If something bad happens, we'd have the stack trace along with the most recent source file and config.

ilyavolodin commented:
I think @btmills suggestion might work. If we would have singleton available that we could populate with the information about the last rule that was executing, then we could use global error handler to catch errors and still retrieve enough information for message to be useful.

alberto commented:
The only catch is this could be a problem if we want to use multiple processes in the future.

ilyavolodin commented:
Ahh.. very good point! Didn't think about that. Hmm... that might be a problem.

platinumazure commented:
@ilyavolodin @alberto That doesn't really have much to do with this PR. It
is true that multiple processes means it will be harder to report errors.
But we would need to aggregate success results in a multiprocess
environment, and we should be able to aggregate errors at that time. From
my perspective, there is already a redesign required once we go parallel. I
don't think that should affect this issue unless we know that
parallelization is about to happen (in which case this single-process
implementation is a waste of time).
On Mar 8, 2016 9:04 AM, "Ilya Volodin" <notifications@github.com> wrote:

> Ahh.. very good point! Didn't think about that. Hmm... that might be a
> problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/eslint/eslint/pull/5483#issuecomment-193816047>.
>

ilyavolodin commented:
@platinumazure Multiple processes discussion was about global error handler with singleton info collector.

nzakas commented:
I'm not sure the perf hit is worth it, either. I'd rather start with areas where we are already throwing errors, as we can just annotate those, and cycle back to things like this. I think we also need a strategy for what we output to the user vs. what we write to a log file. The example in this PR seems more like log file information.

In short, I think we need more discussion around how we want to proceed with better error messaging.

platinumazure commented:
Closing, since it seems the PR was premature. (By the way, having the issue be "accepted" sends a different message than that.)
